### ISSUE TITLE
Training a custom NER doubt

### ISSUE THREAD

**Issue by abinpaul1** [explosion/spaCy#3670]
I really appreciate what you are doing. Thank you.
I saw people asking doubts in the issue section, thats why I decided to ask this doubt here. I went through a lot of issues posted here, but couldn't quite connect the dots completely to my issue. 
Thanks again for helping

I need to train NERs to identify:
(i)Programing languages
(ii)Higher education institute names
(iii)Software technologies
(iv)Course (Example, B.Tech, B.Sc , Masters in Science.....)

for a resume parser

Most of these have a kind of limited type of data. The number of programming languages and technologies is limited. Higher education institutes have Institute,University,School of.. in their names. Similarly courses and software technologies.

How do I train these NERs.
Do I first use entity ruler to create a 'PROGRAMMING LANGUAGE' NER with known languages and then train the EntityRecognizer on an annotated dataset of resumes(not sure of the size I need) to improve the recognizing.

The same for the remaining sets. Do I first add custom made rules,patterns using entity Ruler and then train with the data I create by annotating.
The main problem is that I may not have all the courses and programming languages as part of the annotated training set.
Also while training for my resume, should I annotate the other entities also to avoid "catastrophic forgetting". (Save NER data and then add the custom annotations and then use that to train or if I am using rehearsal.py, how should I use it)

Sorry for so many queries at once.

Thanks again
Abin Paul
abin.paul1@gmail.com

## Your Environment
I am working on google Colab
* **spaCy version:** 2.1.0
* **Platform:** Linux-4.14.79+-x86_64-with-Ubuntu-18.04-bionic
* **Python version:** 3.6.7
* **Models:** en_core_web_lg

**Comment by ines** [comment]
What you're trying to do sounds feasible – but it will require some experimentation to find the best approach.

> Do I first use entity ruler to create a 'PROGRAMMING LANGUAGE' NER with known languages and then train the EntityRecognizer on an annotated dataset of resumes(not sure of the size I need) 

Yes, using rules to create labelled examples faster sounds like a good plan. You can then check the output, correct everything that your rules have missed and then use that as training data for your model. We actually use a very similar approach in our annotation tool [Prodigy](https://prodi.gy) and it's been working well.

Ideally, the goal here is to teach the model to *generalise* – so if the examples you are annotating are representative of what the model will see at runtime, it should also recognise other entities in similar contexts. For example, if the model is trained on a lot of example of "X University" and "University of X" in different contexts, it should ideally also recognize "Y University", even if it's never seen it before.

You probably want to start with a few hundred to a few thousand examples, run your first experiments and then go from there.

> Also while training for my resume, should I annotate the other entities also to avoid "catastrophic forgetting". 

If you also want to keep other entity types that the model predicted previously, then yes, you should include those in your data as well. You can also do this semi-automatically: for each text, get the `doc.ents` and add those entity spans to your training examples.

However, given the number of custom labels you want to train and the fact that they potentially overlap with other existing entity types, you might want to consider training a named entity recognizer from scratch. You won't be able to take advantage of the existing pre-trained model, but you also don't have to spend time "fighting it" and handling arbitrary side-effects of the existing weights. Instead, you can focus on collecting the best possible training data for the entity types you're interested in.
